cmd/tikv-server/src/fork.rs

```diff
@@ -0,0 +1,244 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    fs::ReadDir,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use raft_engine::{env::DefaultFileSystem, Engine as RaftEngine};
+use regex::Regex;
+use tikv::config::TikvConfig;
+
+const SNAP_NAMES: &str = "^.+\\.(meta|sst)$";
+const ROCKSDB_WALS: &str = "^([0-9]+).log$";
+
+/// Remove `path` if it exists, and re-create the directory.
+pub fn remove_and_recreate_dir<P: AsRef<Path>>(path: P) -> Result<(), String> {
+    let path = path.as_ref();
+    if let Err(e) = std::fs::remove_dir_all(path) {
+        if e.kind() != std::io::ErrorKind::NotFound {
+            return Err(format!("remove_dir_all({}): {}", path.display(), e));
+        }
+    }
+    create_dir(path)
+}
+
+/// Symlink snapshot files from the original TiKV instance (specified by
+/// `config`) to `agent_dir`. TiKV may try to change snapshot files, which can
+/// be avoid by setting `raftstore::store::config::snap_apply_copy_symlink` to
+/// `true`.
+pub fn symlink_snaps(config: &TikvConfig, agent_dir: &str) -> Result<(), String> {
+    let mut src = PathBuf::from(&config.storage.data_dir);
+    src.push("snap");
+    let mut dst = PathBuf::from(agent_dir);
+    dst.push("snap");
+    create_dir(&dst)?;
+
+    let ptn = Regex::new(SNAP_NAMES).unwrap();
+    symlink_stuffs(&src, &dst, |name| -> bool { ptn.is_match(name) })
+}
+
+/// Symlink or copy KV engine files from the original TiKV instance (specified
+/// by `config`) to `agent_dir`. Then `agent_dir` can be used to run a new TiKV
+/// instance, without any modifications on the original TiKV data.
+pub fn dup_kv_engine_files(config: &TikvConfig, agent_dir: &str) -> Result<(), String> {
```

Comment:
Why not use checkpoint?

Reply:
What we do is very like `checkpoint` does. However seems `checkpoint` requires
* the db has already been opened, which can be satisfied if the source content is read-only;
* copy instead of symlink if the target path isn't on the same disk, which is too expensive.

Reply:
> the db has already been opened, which can't be satisfied if the source content is read-only

You can try open readonly. I'm not sure if this interface is already ported in rust-rocksdb. Or, you play it safe, only hardlink sst files, and copy everything else.

> copy instead of symlink if the target path isn't on the same disk, which is too expensive.

I don't think you can hardlink file across different disks. And by the way, I didn't notice you are using softlinks before. Why using softlinks? If the original file is deleted, the forked engine will be corrupted.

Reply:
Seems rocksdb doesn't have an option for read-only mode.

> If the original file is deleted, the forked engine will be corrupted.

Yes, so the shadow instance can't be run with the original one concurrently. And it's not necessary for us.

Reply:
> Seems rocksdb doesn't have an option for read-only mode.

There is: `OpenForReadOnly`.

If you think it's too troublesome, you can "only hardlink sst files, and copy everything else", because I don't think it's a good idea to manage your own interpretation of the mutability of RocksDB data files in tikv code.

> so the shadow instance can't be run with the original one concurrently

This approach is adding too much uncertainty to the whole operation. i.e. adding too many "soft limits" that are not enforced at code level. And these loopholes will eventually be encountered given this feature is used frequently enough.

And it's not a real limitation anyway. If you use hardlink then two instances can run concurrently.

Reply:
> Yes, so the shadow instance can't be run with the original one concurrently.

Oh, and I just realized using softlinks would require the original instance to never run again. Because if the shadow runs for a while, then the original runs for a while (not concurrently) and delete some files, then the shadow runs again, it will still report a corruption.